### PR TITLE
[APP] ICR FE Integration

### PR DESCRIPTION
**Pull request by psparacino**
## Description

Carbonmark Integration for ICR credits

**Notes for reviewers:**

1) The retirement data for receipts in dependent on merge of #2048 . This will eventually all be moved to an endpoint.
2) `serialization` is one of three possible query parameters for an individual on the ICR api. `serialization` is used over the other two options (`contractAddress` and ICR's version of a `projectID`) whenever possible for consistency.
3) ICR quantities are regular integers, so the amounts are handled separately for the ICR registry [here](https://github.com/KlimaDAO/klimadao/compare/APP_feature/icr-api-integration?expand=1#diff-1a13cfc031e74278375852de1ca0bf4b8d421b8ed5125f837cd8cc799016d9dbR62). 
4) This PR uses a mainnet testing setup: mainnet RA, carbonmark contract, ICR Carbon Registry contract, ICR subgraph, digital-carbon and marketplace subgraphs. These all need to be changed to production versions before merge.
5) The API is currently hardcoded to the API PR Preview #2051 . This should be [removed](https://github.com/KlimaDAO/klimadao/pull/2052/files#diff-a1be0ef50959b412feee17a3bcc720124c83ccf3e1ab62f240d5ad0d669a1aa4R38) before merge.

To change before merge:
- [x] AggregatorV2 contract
- [x] Carbonmark contract 
- [x] ICR Carbon Registry contract
- [x] ICR subgraph
- [x] Digital-carbon subgraph
- [x] Marketplace subgraph
- [x] Hardcoded Preview URL

Other improvements were noted along the way that weren't necessarily ICR specific but were rolled into this PR:
1) Removing duplicate prices or quantities when only editing one option in a listing: [here](https://github.com/KlimaDAO/klimadao/compare/APP_feature/icr-api-integration?expand=1#diff-7c25eb7f0d621afca61781b09f378ef2b304e011f6bdfe6a1b8785d2b1c616b5R92)
2) Fix recurring nesting error because of title construction in Text component [here](https://github.com/KlimaDAO/klimadao/compare/APP_feature/icr-api-integration?expand=1#diff-4d85183ccd99ed8e794114df235358835b17f2e288ac462bda3c5625383179c0R37)

## Related Ticket

Also related to #2051 

## Notes For QA

Specific pages, components or journeys that might be affected:
- All need to be tested

Relevant preview URLs:
- 

Other notes:
- This is a large PR(!). Please feel free to reach out for any clarification or to pair review any specific section.

### CODE REVIEW

File: carbonmark/lib/constants.ts
Comment (biwano) [inline carbonmark/lib/constants.ts:32]: This does not match with the API version in the API PR

File: carbonmark/lib/createUrls.ts
Comment (biwano) [inline carbonmark/lib/createUrls.ts:15]: this works if `projectData.key.startsWith("ICR") == true` and there is no `projectData.serialization` ?

File: carbonmark/lib/createUrls.ts
Comment (biwano) [inline carbonmark/lib/createUrls.ts:38]: Thank you ser again

**Review by biwano — CHANGES_REQUESTED** [review]

File: carbonmark/components/AssetDetails/index.tsx
Comment (0xemc) [reply carbonmark/components/AssetDetails/index.tsx:63]: Let's nuke ` polyscanUrl: string;` from props now too

File: carbonmark/components/CreateListing/Form/index.tsx
Comment (0xemc) [inline carbonmark/components/CreateListing/Form/index.tsx:104]: Nice pre-emptive catch

File: carbonmark/components/CreateListing/Transaction/index.tsx
Comment (0xemc) [reply carbonmark/components/CreateListing/Transaction/index.tsx:68]: For clarity, this should probably be a labelled formatting function in utils

File: carbonmark/components/CreateListing/index.tsx
Comment (0xemc) [inline carbonmark/components/CreateListing/index.tsx:106]:
```suggestion
const isEIP1155 = !!inputValues?.tokenId;
    // 1155 approvals are all or nothing approvals across all tokenIds and amount
    if (isEIP1155) {
      return currentAllowance === constants.MaxUint256.toString();
    }
```

File: carbonmark/components/pages/Portfolio/Retire/RetireForm/index.tsx
Comment (0xemc) [inline carbonmark/components/pages/Portfolio/Retire/RetireForm/index.tsx:111]: Can you create a ticket for this and link it here please?

File: carbonmark/lib/types/carbonmark.types.ts
Comment (0xemc) [inline carbonmark/lib/types/carbonmark.types.ts:32]: The sooner we integrate with the new retire endpoints and the sooenr we get rid of these types the better.

Do you have a ticket for integrating with the retire endpoints?

File: carbonmark/components/pages/Users/SellerConnected/Forms/EditListing.tsx
Comment (0xemc) [inline carbonmark/components/pages/Users/SellerConnected/Forms/EditListing.tsx:127]: I've seen this in a few places now, can you please extract it as it's own validator.

File: carbonmark/components/pages/Users/SellerConnected/ListingEditable.tsx
Comment (0xemc) [inline carbonmark/components/pages/Users/SellerConnected/ListingEditable.tsx:70]: Is this still relevant? if so is there a ticket we can link to?

File: carbonmark/components/pages/Users/SellerConnected/ListingEditable.tsx
Comment (0xemc) [inline carbonmark/components/pages/Users/SellerConnected/ListingEditable.tsx:118]: As above

File: carbonmark/hooks/useFetchUsersWalletOrHandle.ts
Comment (0xemc) [inline carbonmark/hooks/useFetchUsersWalletOrHandle.ts:1]: This has been resolved with the upgrade to kubb2.0

We can delete this file now.

File: carbonmark/lib/createUrls.ts
Comment (0xemc) [inline carbonmark/lib/createUrls.ts:13]: Are we not able to construct an ICR-1234 id?

File: lib/constants/index.ts
Comment (0xemc) [inline lib/constants/index.ts:78]: Dos this v2 Diamond comment need to go?

**Comment by 0xemc** [comment]
Great effort on this one, it's huge. Some nice clean up along the way as well.

A handful of questions & suggestions then good to go after QA and Bruno approve

File: carbonmark/lib/types/carbonmark.types.ts
Comment (psparacino) [reply carbonmark/lib/types/carbonmark.types.ts:32]: https://github.com/KlimaDAO/klimadao/issues/2128

File: carbonmark/components/pages/Users/SellerConnected/ListingEditable.tsx
Comment (psparacino) [reply carbonmark/components/pages/Users/SellerConnected/ListingEditable.tsx:70]: I'm not sure there's a clear need for it anymore but it'd definitely be a nice to have for the future. Will create a ticket.

File: carbonmark/lib/constants.ts
Comment (psparacino) [reply carbonmark/lib/constants.ts:32]: I had initially understood the MO to be merge first and then open a separate PR that only increments the API version here? Open to whichever option, just want to confirm

File: carbonmark/components/pages/Portfolio/Retire/RetireForm/index.tsx
Comment (psparacino) [reply carbonmark/components/pages/Portfolio/Retire/RetireForm/index.tsx]: Combined ticket to keep track of these items; will break out post-merge: https://github.com/KlimaDAO/klimadao/issues/2093

**Review by 0xemc — APPROVED** [review]
Good to go, 

Obivously wait for the API PR to close and finish it's release first.

Also uncomment the PREVIEW_URL in `constants.ts`

File: carbonmark/lib/constants.ts
Comment (0xemc) [reply carbonmark/lib/constants.ts]: If we merge first then update staging will be broken in that intermediate time right?

Comment (psparacino) [reply carbonmark/lib/constants.ts]: Just tried with the ICR FE pointing at the current api version and everything seemed(*) to work ok except everything ICR obviously.  Everything else being equal it makes sense to me(?) to just include the version incrementation in the APP PR itself.

And just to confirm the versioning:

Currently FE is on 5.0.0.

Currently API is 5.2.1

ICR API is 5.3.1

So the API PRs and APP PRs would increment both FE and API to 5.3.1. (seems like a big jump for the FE so just want to confirm).

**Review by jabby09 — APPROVED** [review]

**Review by biwano — APPROVED** [review]

**Pull request merged**
